sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryStorageStreamSource.java

```diff
@@ -17,161 +17,565 @@
+              rangeTracker.getStartOffset(),
+              splitPosition.getOffset() + 1);
+
+          return rangeTracker.getStopPosition();
+
+        } else {
+
+          BigQueryStorageStreamSource<T> newSource =
+              new BigQueryStorageStreamSource<>(
+                  currentSource.readSession,
+                  response.getRemainderStream(),
+                  splitPosition.getOffset() - response.getPrimaryStream().getRowCount() + 1,
+                  StreamOffsetRangeTracker.OFFSET_INFINITY,
+                  SplitDisposition.SELF,
+                  null,
+                  currentSource.parseFn,
+                  currentSource.outputCoder,
+                  currentSource.bqServices);
+
+          synchronized (this) {
+            this.currentSource = newSource;
+          }
+
+          rangeTracker.resolvePendingSplit(
+              response.getRemainderStream(),
+              splitPosition.getOffset() - response.getPrimaryStream().getRowCount() + 1,
+              StreamOffsetRangeTracker.OFFSET_INFINITY);
+
+          return rangeTracker.getStartPosition();
+        }
+      }
+    }
+
+    private boolean readNextRow() {
+      while (rowIterator == null || !rowIterator.hasNext()) {
         if (!responseIterator.hasNext()) {
-          currentRow = null;
+          responseStream = null;
           return false;
         }
-        rowIterator = responseIterator.next().getRowsList().iterator();
+
+        ReadRowsResponse nextResponse = responseIterator.next();
+        rowIterator = nextResponse.getRowsList().iterator();
+        if (nextResponse.hasStatus()) {
+          estimatedStreamLength = nextResponse.getStatus().getTotalEstimatedRows();
+        }
       }
-      currentRow = rowIterator.next();
+
+      Row nextRow = rowIterator.next();
+      currentRecord = parseFn.apply(new SchemaAndRowProto(schema, nextRow));
       return true;
     }
 
     @Override
     public T getCurrent() {
-      if (currentRow == null) {
-        return null;
-      }
-      return parseFn.apply(new SchemaAndRowProto(readSession.getProjectedSchema(), currentRow));
+      return currentRecord;
     }
 
     @Override
     public void close() {
-      // Do nothing.
+      if (responseStream != null) {
+        responseStream.cancel();
+      }
+      client.close();
+    }
+
+    @Override
+    public Double getFractionConsumed() {
+      return rangeTracker.getOrEstimateFractionConsumed(estimatedStreamLength);
     }
 
     @Override
-    public BoundedSource<T> getCurrentSource() {
-      return source;
+    public synchronized BigQueryStorageStreamSource<T> getCurrentSource() {
+      return currentSource;
+    }
+
+    @Override
+    public synchronized BigQueryStorageStreamSource<T> splitAtFraction(double fraction) {
+      if (currentSource.splitDisposition != SplitDisposition.SELF) {
+        return null;
+      }
+
+      if (currentSource.stopOffset == StreamOffsetRangeTracker.OFFSET_INFINITY) {
```

Comment:
Please clarify when a source will end at StreamOffsetRangeTracker.OFFSET_INFINITY and when it will not.